test/onnx/test_utility_funs.py

```diff
@@ -1097,6 +1097,110 @@ def forward(self, x, y, z):
         for node in graph.nodes():
             self.assertIn(node.scopeName(), expected_torch_script_scope_names)
 
+    def test_scope_of_constants_when_combined_by_cse_pass(self):
+        LAYER_NUM = 3
+
+        class M(torch.nn.Module):
+            def __init__(self, constant):
+                super().__init__()
+                self.constant = constant
+
+            def forward(self, x):
+                # 'self.constant' is designed to be the same for all layers,
+                # hence it is common sub expression.
+                return x + self.constant
+
+        class N(torch.nn.Module):
+            def __init__(self, layers: int = LAYER_NUM):
+                super().__init__()
+                self.layers = torch.nn.ModuleList(
+                    [M(constant=torch.tensor(1.0)) for i in range(layers)]
+                )
+
+            def forward(self, x):
+                for layer in self.layers:
+                    x = layer(x)
+                return x
+
+        graph, _, _ = self._model_to_graph(
+            N(), (torch.randn(2, 3)), input_names=[], dynamic_axes={}
+        )
+
+        # NOTE: Duplicated constants are populated due to implicit casting in scalar_type_analysis,
+        #       so we expect 3 constants with different scopes. The 3 constants are for the 3 layers.
+        #       If CSE in exporter is improved later, this test needs to be updated.
```

Comment:
That means that even if current implementation (CSE enabled as a jit pass for exporter) the constant will be duplicated 3 times instead of being folded to a single constant?

Reply:
Correct. The constants were created by scalar_type_analysis pass. This PR only corrects the meta data of those constants.

Reply:
Why implicit cast is needed, leading to duplicates? would it be preventable?

Reply:
torch supports implicit cast for arithmetic operators and more. ONNX does not. Hence exporter needs to do the casting. 
The duplication can only be prevented by examining the source and uses of the constants, otherwise it wouldn't be safe to delete the original constant, as it might still be used elsewhere. The process is essentially the CSE pass.